### PR TITLE
WIP: Improve idempodency in CloudDataTransferServiceCreateJobOperator

### PR DESCRIPTION
**Pull request by khyurri**
This PR
- adds JOB_NAME to available request body keys
- checks error on creating new transfer

But, I've got 2 open questions: 

1. If user deletes transfer job, new job with same name also raises ALREADY_EXISTS error. 
Right now I just pass result `transfer_job` to caller (line 160 airflow/providers/google/cloud/hooks/cloud_storage_transfer_service.py). `transfer_job["STATUS"] ` will be "DELETED". 

2. `job_name` must be in format "transferJobs/[A-Za-z0-9-._~]*[A-Za-z0-9]$". Should we check it? 

---
Make sure to mark the boxes below before creating PR: [x]

- [ ] Description above provides context of the change
- [ ] Unit tests coverage for changes (not needed for documentation changes)
- [ ] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [ ] Relevant documentation is updated including usage instructions.
- [ ] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by turbaszek** [comment]
> 1. If user deletes transfer job, new job with same name also raises ALREADY_EXISTS error.
>    Right now I just pass result `transfer_job` to caller (line 160 airflow/providers/google/cloud/hooks/cloud_storage_transfer_service.py). `transfer_job["STATUS"] ` will be "DELETED".

In that case we should check the status of job and if it exists but is "DELETED" then we should patch this job using https://cloud.google.com/storage-transfer/docs/reference/rest/v1/transferJobs/patch

So in the case of 409 and status=DELETE, the operator should change the existing job to have "ENABLED" status. 

> 2. `job_name` must be in format "transferJobs/[A-Za-z0-9-._~]*[A-Za-z0-9]$". Should we check it?

It will be validated by API. Having regexp checks on our side requires constant updates (if API changes and our code not we will be misleading users)

**Comment by khyurri** [comment]
> In that case we should check the status of job and if it exists but is "DELETED" then we should patch this job using https://cloud.google.com/storage-transfer/docs/reference/rest/v1/transferJobs/patch
> 
> So in the case of 409 and status=DELETE, the operator should change the existing job to have "ENABLED" status.

I've tried to achieve this, but got error: `Updating a deleted transfer job is not allowed.`

---
From [documentation](https://cloud.google.com/storage-transfer/docs/reference/rest/v1/transferJobs#TransferJob.FIELDS.status):

**DELETED**
This is a soft delete state. After a transfer job is set to this state, the job and all the transfer executions are subject to garbage collection. Transfer jobs become eligible for garbage collection 30 days after their status is set to DELETED.

---

I have not found documentation about `garbage collection` in transferJobs. 

# In conclusion

We cannot create or patch soft deleted jobs before it garbage collected :( 
We can raise exception in this case and mark operator failed. 

Any ideas?

**Comment by turbaszek** [comment]
@khyurri we can try to add unique suffix to job name. In this way, new job will be created anyway. I think nice logging and returning the job name in xcom should help users

**Comment by khyurri** [comment]
@turbaszek 
In this case, if user will not change `JOB_NAME` in `aws_to_gcs_transfer_body` in DAG configuration, we will not achieve  idempodency.

For example, I've got DAG configuration:
```
aws_to_gcs_transfer_body = {
...
JOB_NAME: "transferJobs/helloJobName",
...
}
```

Then I've deleted this transfer job using GCP UI.

Every next dag run will create new job, because `transferJobs/helloJobName` is deleted and we should add suffix to its name.

We can find all jobs and try to check — if this job is matches our format (job_name+suffix). 

But, don't you think, that in this case, when JOB was deleted, right solution is to fail fast?

**Comment by turbaszek** [comment]
> But, don't you think, that in this case, when JOB was deleted, right solution is to fail fast?

As a user, when I run DAG I don't care if someone has deleted the job (for example by mistake), I want it to be created. So, I would suggest adding unique suffix in operator / hook method:

```python
def create_job(job_name, ...):
    job_name = f"{job_name}-{uuid.uuid4()}"
    ...
    return job_name
```

In this way, each job will be unique (up to uuid uniqueness)

**Comment by mik-laj** [comment]
@turbaszek If the job has been completed, do not perform it again. This is a serious problem when user does backfill.  If the user wants to run the task multiple times, the user should not provide any ID and a new ID will be generated each time.
